Add_week-02-01 (#91)
st102/2020-1st-St102-Algorithm-Sprint: week_02/codewars/sylee/200831_Sum_of_Digits.md

# Sum of Digits / Digital Root

## �������

> �ڿ��� n �� �־����, n�� �� �ڸ����� ��� ���Ѵ�. ���� �� �ڸ��� ���� �� ���� ����Ѵ�.

## ���� Ǯ�� ���

 ```python

> def digital_root(n):
>     list_a = [x for x in str(n)]
>     result_a = sum(int(x) for x in list_a)
> 
>     if result_a >= 10:
>         return digital_root(result_a)
>     return result_a

Reviewer: ```suggestion
def digital_root(n):
    list_digit = [digit for digit in str(n)]
    sum_of_digit = sum(int(digit) for digit in list_digit)

    if sum_of_digit >= 10:
        return digital_root(sum_of_digit)
    return sum_of_digit
```

`a` 를 변수명에 쓰는 것보다는 좀 더 의미에 맞도록 써주면 좋을 것 같애~

 ```

> ó��ۼ� �ڵ��̴�.  �� �ڸ��� ����Ʈ�� ��ҷ� ����� ��� ���Ѵ�. 
>
> ��� ��� �Լ� �ٽ� ����ϰų� �ٷ� ������.

```python
def digital_root(n):
    result_a = sum(int(x) for x in str(n))

>     return result_a if result_a < 10 else digital_root(result_a)

Reviewer: 여기도 마찬가지로 `result_a` 라는 변수명에서 `a` 는 무엇인지 전혀 알 수 없는 이름이라서 좀 더 의미있는 이름으로 수정해주면 좋을 것 같애~

```

>� �ڵ忡�� ��� �� �κ�� ����ߴ�. 

1. ���� ����Ʈ�� ����� �ʿ䰡 �ֳ� �;� ��ݴ�.

> 2. if���  �����ϰ� ����ߴ�.

Reviewer: 좋은 리팩토링 포인트를 잡은 것 같애 👍 👍

## �ٸ� ����� Ǯ�� ���

```python
# Best Practices 1
def digital_root(n):

>     return n if n < 10 else digital_root(sum(map(int,str(n))))

Reviewer: 이 풀이에서 중요한 점은 재귀도 중요하지만, 

`sum(map(int, str(n)))` 이걸 잘 기억하고, 나중에 잘 써먹으면 좋을 것 같애~

```

> �� ��� �ڵ忡�� �� �����ϰ� ���� ����̴�.
>
> �۵� ���� �� �ڵ�� ����� �� ����.

```python
# Clever
def digital_root(n):
    return n%9 or n and 9 
```

 > �̰� ������ ȹ������ ����ΰ� ����.

> `return n%9`������ n�� 0�̰ų� 9�� ����� ��츦 ����ϰ� ��� ���� �� �� �� �ִ�.

> `n and 9`������ n�� 0�̸� 0� 9�� ����̸� 9�� ��ȯ�Ѵ�.
>

> > >  `A and B`�� �غ��� A�� B�߿��� ��� ���� ����� �� ����.

Reviewer: 이 풀이는 개인적으로도 좀 놀라운 풀이 였고(사실 완벽히 이해못함.. ㅠ) 투표도 많이 받았지만, 이런 코드들 보다는 좀 더 보편적으로 많이 사용되고, 훨씬 더 읽기 좋은 코드를 보고 배우면 좋을 것 같애~
